contracts/ControllerModule.sol

```diff
@@ -2,71 +2,41 @@
 
 pragma solidity ^0.8.23;
 
-import { RLPReader } from "solidity-rlp/contracts/RLPReader.sol";
-import { MerklePatriciaProofVerifier } from "./libraries/MerklePatriciaProofVerifier.sol";
+import { HashiProver } from "./hashi/HashiProver.sol";
+import { AccountAndStorageProof } from "./hashi/HashiProverStructs.sol";
 import { Enum } from "safe-contracts/contracts/common/Enum.sol";
-import { IShoyuBashi } from "./interfaces/hashi/IShoyuBashi.sol";
 import { ISafe } from "./interfaces/safe/ISafe.sol";
+import { IControllerModule } from "./interfaces/IControllerModule.sol";
 
-contract ControllerModule {
-    using RLPReader for RLPReader.RLPItem;
-    using RLPReader for bytes;
-
-    uint256 internal constant LATEST_COMMITMENTS_SLOT = 0;
+contract ControllerModule is IControllerModule, HashiProver {
+    uint256 public constant LATEST_COMMITMENTS_SLOT = 0;
 
     uint256 public immutable SOURCE_CHAIN_ID;
     address public immutable MAIN_SAFE;
     address public immutable SECONDARY_SAFE;
     address public immutable PERIPHERAL;
-    address public immutable SHOYU_BASHI;
-
-    struct Proof {
-        uint256 blockNumber;
-        uint256 nonce;
-        bytes blockHeader;
-        bytes peripheralAccountProof;
-        bytes peripheralStorageProof;
-    }
-
-    struct SafeTxParams {
-        address to;
-        uint256 value;
-        bytes data;
-        Enum.Operation operation;
-        uint256 safeTxGas;
-        uint256 baseGas;
-        uint256 gasPrice;
-        address gasToken;
-        address payable refundReceiver;
-        bytes signatures;
-    }
-
-    uint256 public expectedNonce;
+    bytes32 public PERIPHERAL_COMMITMENTS_STORAGE_KEY;
 
-    error InvalidLatestPeripheralCommitment(
-        bytes32 latestPeripheralCommitment,
-        bytes32 expectedLatestPeripheralCommitment
-    );
-    error InvalidBlockHeader(bytes32 blockHeaderHash, bytes32 expectedBlockHeaderHash);
-    error InvalidAccountStorageRoot();
-    error InvalidNonce(uint256 nonce, uint256 expectedNonce);
-    error InvalidAccountRlp(bytes accountRlp);
+    uint256 public currentNonce;
 
     constructor(
         uint256 sourceChainId,
         address mainSafe,
         address secondarySafe,
         address peripheral,
         address shoyuBashi
-    ) {
+    ) HashiProver(shoyuBashi) {
         SOURCE_CHAIN_ID = sourceChainId;
         MAIN_SAFE = mainSafe;
         SECONDARY_SAFE = secondarySafe;
         PERIPHERAL = peripheral;
-        SHOYU_BASHI = shoyuBashi;
+        PERIPHERAL_COMMITMENTS_STORAGE_KEY = keccak256(
```

Comment:
Should only has one `keccak256(abi.encode())` instead of two. Otherwise will cause `InvalidStorageKey` error.